yellow-paper/docs/contract-deployment/classes.md

```diff
@@ -193,35 +198,70 @@ The `ContractClassRegisterer` contract also allows broadcasting individual funct
+  private_function_tree_sibling_path: Field[],
+  artifact_function_tree_sibling_path: Field[],
+  function: { selector: Field, metadata_hash: Field, vk_hash: Field, bytecode: Field[] },
 )
   emit_unencrypted_event ClassPrivateFunctionBroadcasted(
     contract_class_id,
-    artifact_metadata,
+    artifact_metadata_hash,
     unconstrained_functions_artifact_tree_root,
-    function_leaf_sibling_path,
+    private_function_tree_sibling_path,
+    artifact_function_tree_sibling_path,
     function,
   )
 ```
 
 ```
 function broadcast_unconstrained_function(
   contract_class_id: Field,
-  artifact_metadata: Field,
+  artifact_metadata_hash: Field,
   private_functions_artifact_tree_root: Field,
-  function_leaf_sibling_path: Field,
-  function: { selector: Field, metadata: Field, bytecode: Field[] }[],
+  artifact_function_tree_sibling_path: Field[],
+  function: { selector: Field, metadata_hash: Field, bytecode: Field[] }[],
 )
   emit_unencrypted_event ClassUnconstrainedFunctionBroadcasted(
     contract_class_id,
-    artifact_metadata,
-    unconstrained_functions_artifact_tree_root,
-    function_leaf_sibling_path: Field,
+    artifact_metadata_hash,
+    private_functions_artifact_tree_root,
+    artifact_function_tree_sibling_path,
     function,
   )
 ```
 
+A node that captures a `ClassPrivateFunctionBroadcasted` should perform the following validation steps before storing the private function information in its database:
+
+```
+// Load contract class from local db
+contract_class = db.get_contract_class(contract_class_id)
+
+// Compute function leaf and assert it belongs to the private functions tree
+function_leaf = pedersen([selector as Field, vk_hash], GENERATOR__FUNCTION_LEAF)
+computed_private_function_tree_root = compute_root(function_leaf, private_function_tree_sibling_path)
+assert computed_private_function_tree_root == contract_class.private_function_root
+
+// Compute artifact leaf and assert it belongs to the artifact
+artifact_function_leaf = sha256(selector, metadata_hash, sha256(bytecode))
+computed_artifact_private_function_tree_root = compute_root(artifact_function_leaf, artifact_function_tree_sibling_path)
+computed_artifact_hash = sha256(computed_artifact_private_function_tree_root, unconstrained_functions_artifact_tree_root, artifact_metadata_hash)
+assert computed_artifact_hash == contract_class.artifact_hash
+```
+
+<!-- TODO: Requiring two sibling paths isn't nice. This is because we are splitting private function information across two trees: one for the protocol, that deals only with selectors and vk hashes, and one for the artifact, which deals with bytecode and metadata. If we are fine adding a `function_stuff_hash` to the function leaf that goes into the protocol tree, we could get rid of the second sibling path, but that introduces stuff into the private function tree that is not strictly needed and requires unnecessary hashing in the kernel. -->
```

Comment:
I reckon aiming to reduce any in-circuit hashing is the best approach, here

Reply:
Sounds good. Moved this to a "discarded approaches" section.